postmates/avro
# Adding avro timestamp to python datetime support

This PR merges in changes originally created at Yelp, with a few modifications.

To test locally:
```
./build.sh docker
cd lang/py
./build.sh test
```

This PR gives us a (partial) bijection* between the following avro types and python types:
* `{'logicalType': 'timestamp-micros', 'type': 'long'}` **<=>** `datetime.datetime`
* `{'logicalType': 'timestamp-millis', 'type': 'long'}` **<=>** `datetime.datetime`
* `{'logicalType': 'time-micros', 'type': 'long'}` **<=>** `datetime.time`
* `{'logicalType': 'time-millis', 'type': 'int'}` **<=>** `datetime.time`
* `{'logicalType': 'date', 'type': 'int'}` **<=>** `datetime.date`
* `{'logicalType': 'decimal', 'type': 'string'}` **<=>** `decimal.Decimal`

I had to make a few modifications to the code originally created by Yelp:
* timezone-aware datetimes were removed, since we use naive datetimes in UTC
* Yelp's changes caused the library to error on unspecified logical types, which goes against the Avro specification. I changed the library to match the spec:
  > Language implementations must ignore unknown logical types when reading, and should use the underlying Avro type. If a logical type is invalid, for example a decimal with scale greater than its precision, then implementations should ignore the logical type and use the underlying Avro type.
http://avro.apache.org/docs/current/spec.html#Logical+Types

Notes on unusual cases:
* **To allow for the possibility that some systems would emit events that contain timestamps already formatted as `int`/`longs`, it's not *precisely* a bijection. Avro schemas with `{'logicalType': 'timestamp-millis', 'type': 'long'}` and `{'logicalType': 'timestamp-micros', 'type': 'long'}` will accept python *long*s or *int*s on write, but will be read as `datetime.datetime`. I discussed this with Dharam, and he agreed it seemed like the right solution.**

* The `{'logicalType': 'decimal', 'type': 'string'}` **<=>** `decimal.Decimal` mapping doesn't match [Avro's bytes/decimal specification](https://avro.apache.org/docs/1.8.1/spec.html#Decimal), but that's okay! Python's `Decimal` class is really just a string with some math functions to avoid floating point error. And logical types are meant to be "FYI" annotations on top of primitive types, to be ignored in favor of the primitive type if a library doesn't know how to handle that logicalType. (I'm also punting on matching the Avro bytes/decimal type because it's the faster thing to do to get this in to prod. Main thing is to get datetime support shipped, everything else is "nice to have.")
